rust-runtime/aws-smithy-checksums/src/lib.rs

```diff
@@ -5,11 +5,111 @@
 
 //! Checksum calculation and verification callbacks.
 
+use ::http::header::HeaderName;
 use bytes::Bytes;
 
 pub mod body;
 pub mod http;
 
+// Valid checksum algorithm names
+pub const CRC_32_NAME: &str = "crc32";
+pub const CRC_32_C_NAME: &str = "crc32c";
+pub const SHA_1_NAME: &str = "sha1";
+pub const SHA_256_NAME: &str = "sha256";
+pub const MD5_NAME: &str = "md5";
+
+/// We only support checksum calculation and validation for these checksum algorithms.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ChecksumAlgorithm {
+    Crc32,
+    Crc32c,
+    Md5,
+    Sha1,
+    Sha256,
+}
+
+impl ChecksumAlgorithm {
+    /// Create a new `ChecksumAlgorithm` from an algorithm name. Valid algorithm names are:
+    /// - "crc32"
+    /// - "crc32c"
+    /// - "sha1"
+    /// - "sha256"
+    /// - "md5"
+    ///
+    /// Passing an invalid name will return an error.
+    pub fn new(checksum_algorithm: &str) -> Result<Self, Error> {
+        if checksum_algorithm.eq_ignore_ascii_case(CRC_32_NAME) {
+            Ok(Self::Crc32)
+        } else if checksum_algorithm.eq_ignore_ascii_case(CRC_32_C_NAME) {
+            Ok(Self::Crc32c)
+        } else if checksum_algorithm.eq_ignore_ascii_case(SHA_1_NAME) {
+            Ok(Self::Sha1)
+        } else if checksum_algorithm.eq_ignore_ascii_case(SHA_256_NAME) {
+            Ok(Self::Sha256)
+        } else if checksum_algorithm.eq_ignore_ascii_case(MD5_NAME) {
+            Ok(Self::Md5)
+        } else {
+            Err(Error::UnknownChecksumAlgorithm(
+                checksum_algorithm.to_owned(),
+            ))
+        }
+    }
+
+    /// Return the `HttpChecksum` implementor for this algorithm
+    pub fn into_impl(self) -> Box<dyn http::HttpChecksum> {
+        match self {
+            Self::Crc32 => Box::new(Crc32::default()),
+            Self::Crc32c => Box::new(Crc32c::default()),
+            Self::Md5 => Box::new(Md5::default()),
+            Self::Sha1 => Box::new(Sha1::default()),
+            Self::Sha256 => Box::new(Sha256::default()),
+        }
+    }
+
+    /// Return the `HeaderName` used to represent this checksum algorithm
+    pub fn into_header_name(self) -> HeaderName {
+        match self {
+            Self::Crc32 => http::CRC_32_HEADER_NAME.clone(),
+            Self::Crc32c => http::CRC_32_C_HEADER_NAME.clone(),
+            Self::Md5 => http::MD5_HEADER_NAME.clone(),
+            Self::Sha1 => http::SHA_1_HEADER_NAME.clone(),
+            Self::Sha256 => http::SHA_256_HEADER_NAME.clone(),
+        }
```

Comment:
as a small stylistic piece, I would probably put this into the HTTP module as a free-function instead of putting it on checksum algorithm—checksum algorithm doesn't have any direct attachment to HTTP